File: blacktie/jatmibroker-xatmi/src/test/java/org/jboss/narayana/blacktie/jatmibroker/core/server/TestSocketServer.java
/*
 * JBoss, Home of Professional Open Source.
 * Copyright 2011, Red Hat, Inc., and individual contributors
 * as indicated by the @author tags. See the copyright.txt file in the
 * distribution for a full listing of individual contributors.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

package org.jboss.narayana.blacktie.jatmibroker.core.server;

import static org.junit.Assert.*;

import java.io.DataOutputStream;
import java.io.IOException;
import java.net.Socket;
import java.util.Properties;

import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;
import org.jboss.narayana.blacktie.jatmibroker.core.transport.Message;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

/**
 * @author zhfeng
 *
 */
public class TestSocketServer {
    private static final Logger log = LogManager.getLogger(TestSocketServer.class);
    private SocketServer server;
    private int port = 12341;
    
    @Before
    public void setUp() throws Exception {
        Properties prop = new Properties();
        prop.setProperty("blacktie.java.socketserver.port", new StringBuffer().append(port).toString());
        server = SocketServer.getInstance(prop);
        log.info("socket server start");
    }
    
    @After
    public void tearDown() throws Exception {
        log.info("socket server stop");
        SocketServer.discardInstance();
    }

    @Test
    public void test() throws Exception {
        server.register(1, null, null);
        Socket socket = new Socket("localhost", port);
        DataOutputStream outs =  new DataOutputStream(socket.getOutputStream());
        int sid = 1;
        int correlationId = 0;
        int rcode = 0;
        int flags = 0;
        int rval = 0;
        int len = 4;
        
        byte[] toSend = new byte[len];
        toSend[0] = 'a';
        String toReplyTo = "(null)";
        String type = "X_OCTET";
        String subtype = "(null)";
        
        send(outs, sid, correlationId, rcode, toSend, flags, rval, toReplyTo, type, subtype);
        send(outs, sid, correlationId, 1, toSend, 1, 1, "test", type, subtype);
        Message msg = server.receiveMessage(sid, 2000);
        assertTrue(msg != null);
        assertTrue(msg.cd == 0);
        assertTrue(msg.type.equals(type));
        assertTrue(msg.len == len);
        assertTrue(msg.rcode == 0);
        assertTrue(msg.flags == 0);
        assertTrue(msg.subtype == null);
        assertTrue(msg.data[0] == toSend[0]);
        assertTrue(msg.rval == 0);
        assertTrue(msg.replyTo == null);
        
        msg = server.receiveMessage(sid, 2000);
        assertTrue(msg != null);
        assertTrue(msg.rcode == 1);
        
        outs.close();
        socket.close();
        server.unregister(1);
    }

    private void send(DataOutputStream outs, int sid, int correlationId, int rcode, byte[] toSend, int flags, int rval, String toReplyTo, String type, String subtype) throws IOException {
        StringBuffer buffer = new StringBuffer();
        buffer.append(sid).append("\n").append(correlationId).append("\n").append(rcode).append("\n").
        append(toSend.length).append("\n").append(flags).append("\n").
        append(rval).append("\n").append(toReplyTo).append("\n").
        append(type).append("\n").append(subtype).append("\n");
        
        int sendlen = buffer.length() + toSend.length;
        outs.writeInt(sendlen);
        
        outs.write(buffer.toString().getBytes(), 0, buffer.length());
        outs.write(toSend, 0, toSend.length);
    }
}

File: blacktie/core/src/test/cpp/TestSocketServer.cxx
/*
 * JBoss, Home of Professional Open Source
 * Copyright 2008, Red Hat, Inc., and others contributors as indicated
 * by the @authors tag. All rights reserved.
 * See the copyright.txt in the distribution for a
 * full listing of individual contributors.
 * This copyrighted material is made available to anyone wishing to use,
 * modify, copy, or redistribute it subject to the terms and conditions
 * of the GNU Lesser General Public License, v. 2.1.
 * This program is distributed in the hope that it will be useful, but WITHOUT A
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
 * You should have received a copy of the GNU Lesser General Public License,
 * v.2.1 along with this distribution; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA  02110-1301, USA.
 */
#include "apr.h"
#include "TestSocketServer.h"
#include "TestAssert.h"
#include "btlogger.h"
#include "apr_strings.h"
#include <stdlib.h>

#define DEF_SOCK_TIMEOUT	(APR_USEC_PER_SEC * 30)

static void messagesAvailableCallback(int bar, bool remove) { }

static void* APR_THREAD_FUNC wait_cb(apr_thread_t *thd, void* data) {
	client_ctx_t* ctx = (client_ctx_t*)data;
	MESSAGE message = { NULL, -1, 0, NULL, NULL, NULL, -1, -1, -1, -1, -1, NULL, NULL,
			false, NULL, NULL, false };

	apr_thread_mutex_lock(ctx->mutex);
	while (ctx->data.empty()) {
		apr_thread_cond_wait(ctx->cond, ctx->mutex);
	}

	if(ctx->data.size() > 0){
		message = ctx->data.front();
		ctx->data.pop();
	}
	apr_thread_mutex_unlock(ctx->mutex);

	btlogger("message.type: %s message.len: %ld message.rcode: %ld",
			  message.type, message.len, message.rcode);

	BT_ASSERT(strcmp(message.type, "X_OCTET") == 0);
	BT_ASSERT(strcmp(message.subtype, "") == 0);
	BT_ASSERT(memcmp(message.data, "test1234", message.len) == 0);
	BT_ASSERT(message.len == 8);
	BT_ASSERT(message.flags == 0);
	BT_ASSERT(message.rcode == 1);
	BT_ASSERT(message.rval == 0);

	free((char*)message.replyto);
	free(message.data);
	free(message.type);
	free(message.subtype);

	apr_thread_exit(thd, APR_SUCCESS);
	return NULL;
}

void TestSocketServer::setUp() {
#ifdef WIN32
	apr_status_t rv = apr_initialize();
	BT_ASSERT(rv == APR_SUCCESS);
#endif
	apr_pool_create(&mp, NULL);
	apr_threadattr_create(&thd_attr, mp);
	server = SocketServer::get_instance(port, messagesAvailableCallback);
	btlogger("start server");
}

void TestSocketServer::tearDown() {
	btlogger("stop server");
	SocketServer::discard_instance();
}

void TestSocketServer::testServer() {
	apr_sockaddr_t *sa;
	apr_socket_t *s;
	apr_status_t rv;
	apr_thread_t *client_thead;

	client_ctx_t* ctx = server->register_client(1, NULL);
	BT_ASSERT(ctx != NULL);

	rv = apr_thread_create(&client_thead, thd_attr, wait_cb, (void*)ctx, mp);
	BT_ASSERT(rv == APR_SUCCESS);

	rv = apr_sockaddr_info_get(&sa, "localhost", APR_INET, port, 0, mp);
	BT_ASSERT(rv == APR_SUCCESS);

	rv = apr_socket_create(&s, sa->family, SOCK_STREAM, APR_PROTO_TCP, mp);
	BT_ASSERT(rv == APR_SUCCESS);

	apr_socket_opt_set(s, APR_SO_NONBLOCK, 1);
	apr_socket_timeout_set(s, DEF_SOCK_TIMEOUT);

	rv = apr_socket_connect(s, sa);
	BT_ASSERT(rv == APR_SUCCESS);

	MESSAGE* msg = (MESSAGE*)apr_pcalloc(mp, sizeof(MESSAGE));
	msg->correlationId = 1;
	msg->flags = 0;
	msg->rcode = 1;
	msg->rval  = 0;
	msg->replyto = NULL;
	msg->type = (char*)"X_OCTET";
	msg->subtype = (char*)"";
	msg->data = (char*)"test1234";
	msg->len = strlen(msg->data);

	char *buf = apr_psprintf(mp, "1\n%ld\n%ld\n%ld\n%ld\n%ld\n%s\n%s\n%s\n",
		(long)msg->correlationId, (long)msg->rcode, (long)msg->len, (long)msg->flags, (long)msg->rval,
		msg->replyto, msg->type, 
		(msg->subtype == NULL || strlen(msg->subtype) == 0) ? "(null)" : msg->subtype);

	apr_size_t len = strlen(buf) + msg->len;
	int sendlen = htonl(len);

	len = sizeof(sendlen);
	apr_socket_send(s, (char*)&sendlen, &len);

	len = strlen(buf);
	apr_socket_send(s, buf, &len);

	len = msg->len;
	apr_socket_send(s, msg->data, &len);

	apr_socket_close(s);

	rv = apr_thread_join(&rv, client_thead);
	BT_ASSERT(rv == APR_SUCCESS);

	server->unregister_client(1);
}
